File: exercises/ddd_infrastructure_layer/typescript/src/application/services/GetDetailsService.ts
import { ApplicationId } from '../../domain/application/ApplicationId';
import type { ApplicationRepository } from '../../domain/application/ApplicationRepository';

export interface GetApplicationDetailsCommand {
  applicationId: string; // UUIDv4
}

export interface ApplicationDetailsResult {
  applicationId: string;
  stageName: string;
  receivedDateTime: string;
  fullName?: string;
  phoneNumber?: string;
  emailAddress?: string;
  motivationLetter?: string;
  statusName?: string;
}

export class GetDetailsService {
  private repo: ApplicationRepository;

  constructor(repo: ApplicationRepository) {
    this.repo = repo;
  }

  async execute(cmd: GetApplicationDetailsCommand): Promise<ApplicationDetailsResult> {
    const id = new ApplicationId(cmd.applicationId);
    const application = await this.repo.getById(id);
    if (!application) {
      throw new Error(`Application not found for id ${cmd.applicationId}`);
    }

    return {
      applicationId: application.id.value,
      receivedDateTime: application.receivedDateTime.value,
      fullName: application.fullName?.value,
      phoneNumber: application.phoneNumber?.value,
      emailAddress: application.emailAddress?.value,
      motivationLetter: application.motivationLetter?.value,
      stageName: application.stage?.name,
      statusName: application.status
    }
  }
}

File: exercises/ddd_infrastructure_layer/java/src/main/java/application/services/GetDetailsService.java
package application.services;

import com.google.inject.Inject;

import domain.application.ApplicationId;
import domain.application.ApplicationRepository;
import domain.application.EmailAddress;
import domain.application.FullName;
import domain.application.MotivationLetter;
import domain.application.PhoneNumber;

import java.time.ZonedDateTime;
import java.util.UUID;

public final class GetDetailsService {
    private final ApplicationRepository repository;

    @Inject
    public GetDetailsService(ApplicationRepository repository) {
        this.repository = repository;
    }

    public GetDetailsResult execute(GetDetailsCommand command) throws Exception {
        ApplicationId id = new ApplicationId(command.applicationId());
        var application = repository.getById(id);
        if (application == null) {
            throw new IllegalStateException("Application not found for id " + command.applicationId());
        }
        FullName fullName = application.getFullName();
        PhoneNumber phoneNumber = application.getPhoneNumber();
        EmailAddress emailAddress = application.getEmailAddress();
        MotivationLetter motivationLetter = application.getMotivationLetter();

        return new GetDetailsResult(
                application.getId().getValue(),
                application.getReceivedDateTime().getValue(),
                fullName != null ? fullName.getValue() : null,
                phoneNumber != null ? phoneNumber.getValue() : null,
                emailAddress != null ? emailAddress.getValue() : null,
                motivationLetter != null ? motivationLetter.getValue() : null,
                application.getStage().getName(),
                application.getStatus().getValue()
        );
    }

    public record GetDetailsCommand(UUID applicationId) {
    }

    public record GetDetailsResult(
            UUID applicationId,
            ZonedDateTime receivedDateTime,
            String fullName,
            String phoneNumber,
            String emailAddress,
            String motivationLetter,
            String stageName,
            String statusName
    ) {
    }
}
